docs/architecture/adr-004-coroutine-effect-system.md

```diff
@@ -28,56 +36,392 @@ We needed a design that would:
+    VoteExtensionValidity(Result<(), VoteExtensionError>),
+}
+```
+
+#### Resumable
+
+The `Resumable` trait allows creating a `Resume` value after having processed an effect.
+
+```rust
+pub trait Resumable<Ctx>
+  where Ctx: Context
+{
+    /// The value type that will be used to resume execution
+    type Value;
+
+    /// Creates the appropriate [`Resume`] value to resume execution with.
+    fn resume_with(self, value: Self::Value) -> Resume<Ctx>;
+}
+```
+
+#### `process!`
+
+The `process!` macro is the entry point for feeding inputs to consensus.
+We omit its implementation here for brevity, but it handles starting the coroutine, processing an input, yielding effects, and resuming consensus with appropriate values.
+
+One can think of it as a function with the following signature, depending on whether the effect handler is synchronous or asynchronous:
+
+```rust
+// If the effect handler is synchronous
+fn process<Ctx>(
+    input: Input<Ctx>,
+    state: &mut ConsensusState<Ctx>,
+    metrics: &Metrics,
+    with: impl FnMut(Effect<Ctx>) -> Result<Resume<Ctx>, Error>,
+) -> Result<(), ConsensusError<Ctx>>
+where
+    Ctx: Context;
+```
+
+```rust
+// If the effect handler is asynchronous
+async fn process<Ctx>(
+    input: Input<Ctx>,
+    state: &mut ConsensusState<Ctx>,
+    metrics: &Metrics,
+    with: impl AsyncFnMut(Effect<Ctx>) -> Result<Resume<Ctx>, Error>,
+) -> Result<(), ConsensusError<Ctx>>
+where
+    Ctx: Context;
+```
+
+### Flow
+
+1. The application calls `process!` with an input, consensus state, metrics, and an effect handler.
+2. This initializes a coroutine which immediately starts processing the input.
+3. The coroutine runs until it needs something from the environment.
+4. At that point, the coroutine yields an `Effect` (like `SignVote` or `GetValue`).
+5. The effect handler performs the requested operation.
+6. For synchronous effects (like `SignVote`), the handler immediately resumes the coroutine with the result.
+7. For asynchronous effects (like `GetValue`), the handler immediately resumes the coroutine with a `()` (unit) value,
+   and will typically schedule a background task to provide the result later by feeding it as a new input back to consensus via the `process!` macro.
+
+#### Possible effects yielded by an input
+
+Here is a list of all effects that can be yielded when processing an input:
+
+* StartHeight:
+  - CancelAllTimeouts
+  - ResetTimeouts
+  - ScheduleTimeout
+  - GetValue
+  - Effects for any pending inputs
+
+* Vote
+  - WalAppendMessage
+  - VerifyVoteExtension
+  - GetValidatorSet
+  - VerifySignature
+
+* Proposal:
+  - WalAppendMessage
+  - GetValidatorSet
+  - VerifySignature
+  - Publish
+  - Same as *Vote*
+
+* Propose:
+  - RestreamProposal
+  - WalAppendMessage
+  - GetValidatorSet
+  - VerifySignature
+  - Publish
+
+* TimeoutElapsed;
+  - WalAppendTimeout
+  - Decide
+
+* ProposedValue:
+  - CancelTimeout
+  - Same as *Vote*
```

Comment:
Vote or Proposal?

Reply:
Vote, as this may trigger a vote which will be fed back to consensus straight away